File: video-renderer/video_summary.py
"""Summary layout intent, kept in sync with server/videoSummary.ts."""
import re


def is_summary_heading(text):
    return isinstance(text, str) and bool(re.search(r'总结|小结|回顾|复盘|\b(?:summary|recap)\b', text, re.I | re.ASCII))


def resolve_shot_template(shot):
    template = (shot.get('layout') or {}).get('template', 'auto')
    if template == 'summary' or is_summary_heading(shot.get('title')) or is_summary_heading(shot.get('sectionTitle')):
        return 'summary'
    return template if template not in ['auto', 'question'] else 'explain'


def summary_shot_ids(project):
    indices = {line['id']: index for index, line in enumerate(project.get('utterances', []))}
    boundaries = []
    for order, note in enumerate(project.get('scriptNotes') or []):
        if note.get('kind') not in ['chapter', 'summary'] or note.get('utteranceId') not in indices:
            continue
        index = indices[note['utteranceId']] + (1 if note.get('placement') == 'after' else 0)
        boundaries.append((index, note.get('sourceLine', 0), order,
                           note['kind'] == 'summary' or is_summary_heading(note.get('text'))))
    boundaries.sort(key=lambda value: value[:3])
    result, boundary, structural_summary, heading_section = set(), 0, False, None
    for shot in project['shots']:
        first = next((indices[uid] for uid in shot.get('utteranceIds', []) if uid in indices), None)
        while first is not None and boundary < len(boundaries) and boundaries[boundary][0] <= first:
            structural_summary = boundaries[boundary][3]
            boundary += 1
            heading_section = None
        section_title = (shot.get('sectionTitle') or '').strip()
        if heading_section is not None and (shot.get('chapter') != heading_section[0] or
                section_title and heading_section[1] and section_title != heading_section[1]):
            heading_section = None
        if is_summary_heading(shot.get('title')) or is_summary_heading(shot.get('sectionTitle')):
            heading_section = (shot.get('chapter'), section_title)
        if (shot.get('layout') or {}).get('template') == 'summary' or structural_summary or heading_section is not None:
            result.add(shot['id'])
    return result


def normalize_summary_project(project):
    ids = summary_shot_ids(project)
    return {**project, 'shots': [{**shot, 'layout': {**(shot.get('layout') or {}),
            'template': 'summary', 'elements': (shot.get('layout') or {}).get('elements') or {}}}
            if shot['id'] in ids else shot for shot in project['shots']]}

File: server/videoSummary.ts
import type { Shot, VideoProject } from './videoTypes.js'

/** Only production headings are inspected; spoken dialogue and review prose are not headings. */
export function isVideoSummaryHeading(text: string | undefined): boolean {
  return typeof text === 'string' && /总结|小结|回顾|复盘|\b(?:summary|recap)\b/i.test(text)
}

type SummaryProject = Pick<VideoProject, 'shots' | 'utterances' | 'scriptNotes'>

/** Resolve the full ordered project before selecting a single page for preview or rendering. */
export function summaryShotIds(project: SummaryProject): Set<string> {
  const indices = new Map(project.utterances.map((line, index) => [line.id, index]))
  const boundaries = (project.scriptNotes || []).flatMap((note, order) => {
    if (note.kind !== 'chapter' && note.kind !== 'summary') return []
    const anchor = note.utteranceId ? indices.get(note.utteranceId) : undefined
    if (anchor === undefined) return []
    const index = anchor + (note.placement === 'after' ? 1 : 0)
    return [{ index, order, sourceLine: note.sourceLine, summary: note.kind === 'summary' || isVideoSummaryHeading(note.text) }]
  }).sort((a, b) => a.index - b.index || a.sourceLine - b.sourceLine || a.order - b.order)
  const result = new Set<string>()
  let boundary = 0, structuralSummary = false
  let headingSection: Pick<Shot, 'chapter' | 'sectionTitle'> | undefined
  for (const shot of project.shots) {
    const first = shot.utteranceIds.map(id => indices.get(id)).find(index => index !== undefined)
    // A structural heading starts a fresh section even when an AI reused its chapter number.
    while (first !== undefined && boundary < boundaries.length && boundaries[boundary].index <= first) {
      structuralSummary = boundaries[boundary++].summary
      headingSection = undefined
    }
    if (shot.story) { headingSection = undefined; continue }
    if (headingSection && (shot.chapter !== headingSection.chapter ||
      shot.sectionTitle?.trim() && headingSection.sectionTitle?.trim() && shot.sectionTitle.trim() !== headingSection.sectionTitle.trim())) headingSection = undefined
    if (isVideoSummaryHeading(shot.title) || isVideoSummaryHeading(shot.sectionTitle)) {
      headingSection = { chapter: shot.chapter, sectionTitle: shot.sectionTitle }
    }
    if (shot.layout?.template === 'summary' || structuralSummary || headingSection) result.add(shot.id)
  }
  return result
}

/** Structural summary intent takes precedence over an older automatically chosen page template. */
export function normalizeVideoSummaryLayouts<T extends SummaryProject>(project: T): T {
  const ids = summaryShotIds(project)
  let changed = false
  const shots = project.shots.map(shot => {
    if (!ids.has(shot.id) || shot.layout?.template === 'summary') return shot
    changed = true
    return { ...shot, layout: { ...shot.layout, template: 'summary' as const, elements: shot.layout?.elements || {} } }
  })
  return changed ? { ...project, shots } : project
}
